docs/source/notes/extending.rst

```diff
@@ -52,6 +52,9 @@ encode the operation history. Every new function requires you to implement 2 met
     mark any input that is modified inplace by the forward function.
   - :meth:`~torch.autograd.function._ContextMethodMixin.mark_non_differentiable` must
     be used to tell the engine if an output is not differentiable.
+  - :meth:`~torch.autograd.function._ContextMethodMixin.set_materialize_grads` must
```

Comment:
This documentation isn't clear enough.  The term "undefined output grad tensors" is never defined (hehe), and in fact an "undefined tensor" is not even a concept in python, so referring to it in python documentation doesn't make sense.

I'd suggest something along the lines of:
"the autograd engine is capable of optimizing gradient computations in cases where the output does not depend on the input.  Instead of materializing a tensor filled with zeros, a `None` object (in python) or "undefined tensor" (tensor x for which `x.defined()` is False) in C++.  However, supporting this optimization means your custom autograd function has to handle gradients that are represented in this way and is thus opt-in..."

I'd also make a separate example demonstrating only this -- the example is complicated enough already.

Reply:
Got it, I'll update the doc. Thanks!